chichiwang/exercise-js
# Replace `app/Routes` with a data router `app/dataRoutes`

## Description
Linked to Issue: #19
The only clear way for the Express server [to provide the correct status codes](https://reactrouter.com/en/main/guides/ssr#redirects) using [React-Router](https://reactrouter.com/) is to use a [data router](https://reactrouter.com/en/main/guides/ssr#with-a-data-router). In order to make our webserver compliant with correct status code usage, the component router `app/Routes` needs to be converted to a data router.

This PR makes the change, which is large: this change impacts the entire application from the component structure of the root `<App />` component, to how the client and server entry files render the app. In addition to this change, this PR also ensures that any status codes provided by the router are reflected in the server Response, as well as ensuring that 404 page renders return the correct status code.

## Changes
* Replace `app/Routes` with `app/dataRoutes`
  * An exported data router to be used on both the client and server for routing within the App
* Refactor `app/App` to accept a child component
  * Responsibility for instrumenting the data router belongs to `client/entry` and `server/appHandler` which will render it as a child of `<App />`
* Refactor `client/entry`
  * Create a [BrowserRouter](https://reactrouter.com/en/main/router-components/browser-router) feeding it the data routes
  * Render the `BrowserRouter` as a child of `App` when hydrating root
* Refactor `server/index`
  * Move the app handler, passed as callback to `app.get` on the route `*` to its own module `server/appHandler`
* Create a `server/createFetchRequest` utility to convert Express Request and Response objects to a Fetch Request
  * Borrowed directly from the [React Router documentation](https://reactrouter.com/en/main/guides/ssr#with-a-data-router) on server side rendering
    * An attempt was made to locate a third party, maintained utility to do this, but one was not found. A note in the form of a comment has been made within the module.
* Create a new module `server/appHandler` to house the old `app.get` handler on the server for App routing and response rendering
  * Refactor the handler to create a [StaticRouter](https://reactrouter.com/en/main/routers/create-static-router#createstaticrouter) using the data router and to render it as a child of the `<App />` component
  * Instrument checks for requests that match the route path `*` and apply a status code of 404 to the Response in those cases
  * In non-404 cases, the Response object status code will be set to whatever is provided by the data router
* `.eslintrc.js` has been updated to disable rule `no-restricted-syntex` to allow the use of `For .. In` loops
  * Used by the implementation of `server/createFetchRequest` as borrowed from React-Router's documentation
* `jest-config/paths` has been updated to point to the new server entry file `server/index.ts` (previously `server/index.tsx`)
* Update all test files to reflect and ensure these changes

## Steps to QA
* Pull down and switch to this branch
* Run `npm start` to ensure the assets build and the watch server starts correctly
* Navigate to the server in browser and ensure the home page renders correctly
  * Open the developer tools network tab and hard refresh here to ensure the response from the server has a 200 status code
* Click on the link to the Counter page and use the back button to ensure the routing functionality is not broken
* Navigate to a non-existent route (ex: `/foo/bar/baz`) on the server and ensure the Page Not Found page renders
  * Open the developer tools network tab and hard refresh here to ensure the response from the server has a 404 status code
